dune/grid/polyhedralgrid/grid.hh

```diff
@@ -0,0 +1,1207 @@
+          return 1;
+      }
+      return 0;
+    }
+
+    template <class EntitySeed>
+    GlobalCoordinate
+    corner( const EntitySeed& seed, const int i ) const
+    {
+      const int codim = EntitySeed :: codimension;
+      switch (codim)
+      {
+        case 0:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * cellVertices_[ seed.index() ][ i ];
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+        case 1:
+        {
+          const int faceVertex = grid_.face_nodes[grid_.face_nodepos[seed.index()] + i];
+          return copyToGlobalCoordinate( grid_.node_coordinates + GlobalCoordinate :: dimension * faceVertex );
+        }
+        case dim:
+          {
+            const int coordIndex = GlobalCoordinate :: dimension * seed.index();
+            return copyToGlobalCoordinate( grid_.node_coordinates + coordIndex );
+          }
+      }
+      return GlobalCoordinate( 0 );
+    }
+
+    template <class EntitySeed>
+    int subEntities( const EntitySeed& seed, const int codim ) const
+    {
+      const int index = seed.index();
+      switch (codim)
+      {
+        case 0:
+          return 1;
+        case 1:
+          return grid_.cell_facepos[ index+1 ] - grid_.cell_facepos[ index ];
+        case dim:
+          return cellVertices_[ index ].size();
+      }
+      return 0;
+    }
+
+    template <int codim>
+    typename Codim<codim>::EntitySeed
+    subEntitySeed( const typename Codim<0>::EntitySeed& elemSeed, const int i ) const
+    {
+      assert( i>= 0 && i<subEntities( elemSeed, codim ) );
+      typedef typename Codim<codim>::EntitySeed  EntitySeed;
+      if( codim == 0 )
+        return EntitySeed( elemSeed.index() );
+      else if ( codim == 1 )
+      {
+        return EntitySeed( grid_.cell_faces[ grid_.cell_facepos[ elemSeed.index() ] + i ] );
+      }
+      else if ( codim == dim )
+      {
+        return EntitySeed( cellVertices_[ elemSeed.index() ][ i ] );
+      }
+      else
+      {
+        DUNE_THROW(InvalidStateException,"codimension not availalbe");
+      }
+    }
+
+    const std::vector< GeometryType > &geomTypes ( int codim ) const
+    {
+      static std::vector< GeometryType > emptyDummy;
+      if (0 <= codim && codim < geomTypes_.size())
+        return geomTypes_[codim];
+
+      return emptyDummy;
+    }
+
+    int indexInInside( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      if( ! grid_.cell_facetag )
+        return i;
+      else
+      {
+        // assert( i>= 0 && i<subEntities( EntitySeed( seed.index() ) ) );
+        return grid_.cell_facetag[ grid_.cell_facepos[ seed.index() ] + i ] ;
+      }
+    }
+
+    typename Codim<0>::EntitySeed
+    neighbor( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      typedef typename Codim<0>::EntitySeed EntitySeed;
+      const int face = 2 * this->template subEntitySeed<1>( seed, i ).index();
```

Comment:
This is misleading. The face index is what you have before you multiply by 2. I suggest moving the multiplication to lines 897 and 899, then it will look like tons of other code using the UnstructuredGrid. Also: braces for the `if` below.